pkg/controller/cdapmaster/cdapmaster_controller.go

```diff
@@ -353,6 +395,58 @@ func logbackFromFile(t string, data interface{}) (string, error) {
 	return output.String(), nil
 }
 
+func SetStatusFailed(master *alpha1.CDAPMaster, message string) {
+	master.Status.UpgradeJobStartTimeNano = 0
+	master.Status.Meta.SetCondition(upgradeFailed, message, message)
+}
+
+// Adds a component list object in state in progress. This will cause UpdateStatus to update the
+// ready type to state "false".
+func AddUpgradeComponentNotReady(master *alpha1.CDAPMaster) {
+	os := status.ObjectStatus{}
+	os.Name = "upgrade"
+	os.Status = status.StatusInProgress
+	master.Status.ComponentMeta.ComponentList.Objects = append(master.Status.ComponentMeta.ComponentList.Objects, os)
+}
+
+func LaunchJob(master *alpha1.CDAPMaster, rsrclabels map[string]string) ([]reconciler.Object, error) {
```

Comment:
BTW, it is not actually launching right? The name seems a bit misleading. Shouldn't this just merge with the `GetJobResources` below?

Reply:
This sets/clears additional status fields when the job is first launched